src/node.rs

```diff
@@ -45,19 +45,28 @@ impl Node {
     }
 
     pub fn update_height_size(&mut self) {
-        let left = self.left.as_ref().unwrap();
-        let right = self.right.as_ref().unwrap();
-        self.height = cmp::max(left.height, right.height) + 1;
-        self.size = left.size + right.size;
+        // We add 1 to the child node's height because empty subtrees are assigned a height of -1,
```

Comment:
`update_height_size` and `balance_factor` should never be called on a leaf node, I think it should simply panic in that case.